buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/DistributedPerformanceReporter.groovy

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.testing
+
+import com.google.common.annotations.VisibleForTesting
+import groovy.json.JsonSlurper
+import groovy.transform.TypeChecked
+import groovy.transform.TypeCheckingMode
+import org.gradle.api.internal.tasks.testing.junit.result.TestClassResult
+import org.gradle.api.internal.tasks.testing.junit.result.TestResultSerializer
+
+import java.util.concurrent.atomic.AtomicLong
+
+class DistributedPerformanceReporter extends PerformanceReporter {
+    DistributedPerformanceReporter(PerformanceTest performanceTest) {
+        super(performanceTest)
+    }
+
+    private DistributedPerformanceTest getDistributedPerformanceTest() {
+        return (DistributedPerformanceTest) performanceTest
+    }
+
+    @Override
+    void report() {
+        if (isGeneratingScenarioList()) {
+            // do nothing
+        } else if (!distributedPerformanceTest.isRerun()) {
+            // first run, only write report when it succeeds
+            if (allWorkerBuildsAreSuccessful()) {
+                super.report()
```

Comment:
I wonder if we could use a `delegate` reporter instead of extending the `PerformanceReporter` and calling `super.`. The `DistributedPerformanceReporter` doesn't do much reporting anyway - it is writing and reading some test results. Maybe it would be better if it would only control an actual reporter, without being or extending one.

Reply:
It actually is. The `report()` is mostly doing a `javaexec`, both `DistributedReporter` and `Reporter` need it. The other problems are the fields configured so I'd prefer inheritance here.

Reply:
I am more and more convinced that the handling when to call the reporter and generating the JSON/output files should go into `DistributedPerformanceTest` and has nothing to do with the reporter itself.